### DISCUSSION TITLE
怎么在不修改代码的情况下设置so导出符号

### DISCUSSION THREAD

**Discussion by Elite-stay** [xmake-io/xmake#4959] [Q&A]
release下试了`set_symbols("debug")`、`set_symbols("none")`、`add_rules("utils.symbols.export_list")`都没效果，都是只有额外的sym能导出符号，但是我希望是so本身有

**Comment by Elite-stay** [comment]
另外有没有只导出其中一个链接库的符号的，比如libA.so链接了现成的libB.so，`add_links`里有没有什么特殊的语法可以在构建A的时候导出B包含的符号

**Reply by waruqi** [reply]
默认情况下，so 就是 default visibility ，也就是默认就是全部符号导出的，不需要任何配置。。

xmake 的配置，如果你配置了 `add_rules("mode.release")`，那么 release 编译，默认才会给你禁用符号。。切到 debug ，或者 `set_symbols("none")` 就行了

> release下试了set_symbols("debug")、set_symbols("none") add_rules("utils.symbols.export_list")都没效果，都是只有额外的sym能导出符号

肯定有效果，我不知道你说的没效果是啥，`nm ` 自己确认下符号。

> 另外有没有只导出其中一个链接库的符号的，比如libA.so链接了现成的libB.so，add_links里有没有什么特殊的语法可以在构建A的时候导出B包含的符号

建议你先了解下 so 的基础概念, 别把 so 和 静态库搞混了，A.so 不可能导出 B.so 的

**Reply by Elite-stay** [reply]
不好意思问错了，B是 .a

是配置了`add_rules("mode.release")`，然后没有效果就是，这三个分别试过用`nm`测都是无符号，函数本身是没有写`visiable`的，就是不想修改代码，因为有很多以前用`def`的工程要迁移

**Comment by waruqi** [comment]
> 不好意思问错了，B是 .a
> 
> 是配置了`add_rules("mode.release")`，然后没有效果就是，这三个分别试过用`nm`测都是无符号，函数本身是没有写`visiable`的，就是不想修改代码，因为有很多以前用`def`的工程要迁移

先确认下 debug mode 是否会导出，或者移除 mode.release rules

**Reply by Elite-stay** [reply]
`mode.debug`下是可以，`mode.release`就不行了，`mode.releasedbg`也可以，`mode.release`下`set_symbols("none")`也没效果，这个我试了新建一个`qt.shared`库也是一样的，`-vD`也没看出来，生成cmake的话倒是就算是`release`，也能导出符号

另外提的导出 .a 的符号找到解决方案`add_linkgroups`了

**Reply by waruqi** [reply]
既然 debug 可以，那不就是 set_symbols/set_strip 的问题么，你自己配下就行了。。

自己对比下 https://github.com/xmake-io/xmake/blob/7e7b37e4f776e76d9a88b38f6419ff5c95923ef4/xmake/rules/mode/xmake.lua#L49

> 生成cmake的话倒是就算是release，也能导出符号

不要去对比这个，没啥意义，xmake 的 release 要尽可能小，符号导出都是建议自己代码里面自己导出的，这样生成的 so 能尽可能最小最优，而不是默认一股脑全导出。。

https://github.com/xmake-io/xmake/blob/7e7b37e4f776e76d9a88b38f6419ff5c95923ef4/tests/projects/c/shared_library/src/foo.h#L8

当然，你也可以不用内置的 mode.release 。。。自己配 mode

**Comment by waruqi** [comment]
另外，对于 so ，目前 release mode 默认就是会导出的。。

https://github.com/xmake-io/xmake/blob/7e7b37e4f776e76d9a88b38f6419ff5c95923ef4/xmake/rules/mode/xmake.lua#L48

你这里不行，我估计你根本没说清楚你这里的配置，我猜你是 shared target 去 link 了一个 static target，而你的导出符号是在这个 static target 里，而不是 so 所在的 target。。

这种情况，你得对这个 static target 去 set_symbols none，而不是对 shared target，否则肯定不生效。。符号在哪个 target ，就对哪个 target 配置

**Comment by Elite-stay** [comment]
> 你这里不行，我估计你根本没说清楚你这里的配置，我猜你是 shared target 去 link 了一个 static target，而你的导出符号是在这个 static target 里，而不是 so 所在的 target。。

但是我这编出来就是啥都没有耶，纯的qt.shared

[test.zip](https://github.com/xmake-io/xmake/files/14965357/test.zip)

```shell
$ nm build/linux/x86_64/release/libtest.so 
nm: build/linux/x86_64/release/libtest.so: no symbols
```

```shell
$ xmake build -rvD
[ 50%]: cache compiling.release src/demo.cpp
/usr/bin/gcc -c -m64 -O3 -std=c++11 -I/home/administrator/test/build/.gens/test/linux/x86_64/release/rules/qt/ui -DQT_DEMO_LIBRARY -DQT_NO_DEBUG -DQT_DEPRECATED_WARNINGS -DQT_GUI_LIB -DQT_CORE_LIB -isystem /home/administrator/Qt5.14.2/5.14.2/gcc_64/include/QtGui -isystem /home/administrator/Qt5.14.2/5.14.2/gcc_64/include/QtGui/5.14.2 -isystem /home/administrator/Qt5.14.2/5.14.2/gcc_64/include/QtGui/5.14.2/QtGui -isystem /home/administrator/Qt5.14.2/5.14.2/gcc_64/include/QtCore -isystem /home/administrator/Qt5.14.2/5.14.2/gcc_64/include/QtCore/5.14.2 -isystem /home/administrator/Qt5.14.2/5.14.2/gcc_64/include/QtCore/5.14.2/QtCore -isystem /home/administrator/Qt5.14.2/5.14.2/gcc_64/include -isystem /home/administrator/Qt5.14.2/5.14.2/gcc_64/mkspecs/linux-g++ -DNDEBUG -fPIC -o build/.objs/test/linux/x86_64/release/src/demo.cpp.o src/demo.cpp
[ 75%]: linking.release libtest.so
/usr/bin/g++ -o build/linux/x86_64/release/libtest.so build/.objs/test/linux/x86_64/release/src/demo.cpp.o -shared -m64 -fPIC -L/home/administrator/Qt5.14.2/5.14.2/gcc_64/lib -Wl,-rpath=/home/administrator/Qt5.14.2/5.14.2/gcc_64/lib -s -lQt5Gui -lQt5Core

build cache stats:
cache directory: build/.build_cache
cache hit rate: 100%
cache hit: 1
cache hit total time: 0.000s
cache miss: 0
cache miss total time: 0.000s
new cached files: 0
remote cache hit: 0
remote new cached files: 0
preprocess failed: 0
compile fallback count: 0
compile total time: 0.000s

[100%]: build ok, spent 0.275s
```

**Reply by waruqi** [reply]
不是有的么，哪不行了？

```console
$ nm -D build/linux/x86_64/release/libtest.so 
                 w _ITM_deregisterTMCloneTable
                 w _ITM_registerTMCloneTable
0000000000001110 T _Z1fv
0000000000001100 T _ZN6QtDemoC1Ev
0000000000001100 T _ZN6QtDemoC2Ev
                 w __cxa_finalize
                 w __gmon_start__
                 U qt_version_tag@Qt_5.15
```

**Reply by Elite-stay** [reply]
好吧，好像是可以，不知道要加-D
